docs/switch.rst

```diff
@@ -0,0 +1,369 @@
+=====================================
+Switching from GDAL's Python bindings
+=====================================
+
+Switching from GDAL's Python bindings, meaning the ``gdal`` module in the
+``osgeo`` namespace, to Rasterio for new projects or new iterations on existing
+projects may not be complicated. This document explains the key similarities
+and differences between these two Python packages and highlights the features
+of Rasterio that can help in switching.
+
+Mutual Incompatibility
+======================
+
+Rasterio and GDAL's bindings contend for global GDAL objects. Choose one of
+``import osgeo.gdal`` or ``import rasterio``.
+
+GDAL's bindings (``gdal`` for the rest of this document) and Rasterio are
+mutually incompatible and should not be imported and used in a single Python
+program. The reason is that the dynamic library they each load (these are
+C extension modules, remember), ``libgdal.so`` on Linux, ``gdal.dll`` on
+Windows, has a number of global objects and the two modules take different,
+naive approaches to managing these objects. It is assumed in multiple places in
+``gdal`` that no other Python module is modifying the variables in the dynamic
+library. Rasterio makes the same kind of assumptions and consequently the state
+of global objects like the format driver registry may be unpredictable.
+Additionally, ``gdal`` and Rasterio register conflicting error handlers and
+thus the propagation of exceptions and warnings may depend on which module was
+imported last.
+
+Static linking of the GDAL library for ``gdal`` and ``rasterio`` can avoid
+this contention, but in practice you will almost never see distributions of
+these modules that statically link GDAL.
+
+Beyond the issues above, the modules have different styles and don't complement
+each other well.
```

Comment:
I'd prefer to substantiate or explain this with some examples. Maybe just add that

> While the  `gdal` API directly mirrors that of the libgdal C API, Rasterio's API is designed around idioms found in common Python libraries.

Or something similar.